packages/apollo-gateway/src/__tests__/gateway/lifecycle-hooks.test.ts

```diff
@@ -0,0 +1,153 @@
+                Array [
+                  [GraphQLError: [product] Book -> \`Book\` is an extension type, but \`Book\` is not defined in any service],
+                ]
+            `);
+
+      expect(experimental_didFailComposition).toBeCalled();
+      done();
+    }
+  });
+
+  it('calls experimental_didUpdateComposition on schema update', async done => {
+    const update = jest.fn();
+    update.mockImplementationOnce(async (config: GatewayConfig) => {
+      return [serviceDefinitions, true];
+    });
+    update.mockImplementationOnce(async (config: GatewayConfig) => {
+      const services = serviceDefinitions.filter(s => s.name !== 'books');
+      // overwrite books service with a similar 'book' service
+      return [
+        [
+          ...services,
+          {
+            name: 'book',
+            typeDefs: books.typeDefs,
+            url: 'http://localhost:32542',
+          },
+        ],
+        true,
+      ];
+    });
+
+    const experimental_didUpdateComposition = jest.fn();
+
+    const gateway = new ApolloGateway({
+      experimental_updateServiceDefinitions: update,
+      experimental_pollInterval: 10,
+      experimental_didUpdateComposition,
+    });
+
+    const { interval } = await gateway.load();
+    await new Promise(resolve => setTimeout(resolve, 20));
+
+    clearInterval(interval);
+    const {
+      calls: [firstCall, secondCall],
+    } = experimental_didUpdateComposition.mock;
+
+    expect(experimental_didUpdateComposition).toHaveBeenCalledTimes(2);
+
+    // first call's `current` arg should have a schema
+    expect(firstCall[0].schema).toBeDefined();
+    expect(firstCall[1]).toBeUndefined();
+
+    // second call should have a previous schema
+    expect(secondCall[0].schema).toBeDefined();
+    expect(secondCall[1].schema).toBeDefined();
+    done();
+  });
+
+  it('uses default service definition updater', async () => {
+    const gateway = new ApolloGateway({
+      localServiceList: serviceDefinitions,
+    });
+
+    const { schema } = await gateway.load();
+
+    // spying on gateway.loadServiceDefinitions wasn't working, so this also
+    // should test functionality. If there's no overwriting service definition
+    // updater, it has to use the default. If there's a valid schema, then
+    // the loader had to have been called.
+    expect(schema.getType('User')).toBeDefined();
+  });
+
+  it('warns when polling on the default fetcher', async () => {
+    const consoleSpy = jest.spyOn(console, 'warn');
+    const gateway = new ApolloGateway({
+      serviceList: serviceDefinitions,
+      experimental_pollInterval: 10,
+    });
+    expect(consoleSpy).toHaveBeenCalledTimes(1);
+    expect(consoleSpy.mock.calls[0]).toMatchInlineSnapshot(`
+      Array [
+        "Polling running services is dangerous and not recommended in production. Polling should only be used against a registry. If you are polling running services, use with caution.",
```

Comment:
I like this log line btw (I saw it earlier when running the gateway), and I think that we could probably provide more guidance / reasoning around why. Ideally, we could point to the docs, but unfortunately, the only docs that we have for this are docs for our commercial-ish free-ish product. I think we can link to those here, but we should probably have more open-source style docs that have ideas around why you don't want to poll running services. Also it seems like this functionality is literally turned on by default if you just pass a serviceList, but I could be wrong. If so, that's definitely a bug, no?

Reply:
@zionts Do we think it's a problem to see this warning in dev? Maybe we could check the node env to see if it's in dev and suppress it there? Do you think that'd be good enough?

regarding the message, that was the original intention. We wanted to link out to docs, but I figured it'd be fine without a link until we have one. Should we link to commercial products until we have something better?